frontends/onnx-frontend/onnx-frontend/src/Conversion/OFInsertNecessaryCast.cpp

```diff
@@ -0,0 +1,104 @@
+//===- OFInsertNecessaryCast.cpp
+//-------------------------------------------===//
+//
+// Copyright 2022 ByteDance Ltd. and/or its affiliates. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+
+#include "onnx-frontend/src/Conversion/OFInsertNecessaryCast.hpp"
+
+#include "third_party/onnx-mlir/src/Dialect/ONNX/DialectBuilder.hpp"
+
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+namespace {
+//===----------------------------------------------------------------------===//
+// Unsqueeze
+//===----------------------------------------------------------------------===//
+
+// This pattern insert cast op to force convert unsqueeze data value to the
+// required element type.
+// clang-format off
+// from:
+// %unq = onnx.Unsqueeze(%val, %axes) : tensor<?xTy0> -> tensor<?xTy0>
+// onnx.Concat(..., %unq, ...) -> tensor<?xTy1>
+// to:
+// %cast = onnx.Cast(%val) : tensor<?xTy0> -> tensor<?xTy1>
+// %unq = onnx.Unsqueeze(%cast, %axes) : tensor<?xTy1> -> tensor<?xTy1>
+// onnx.Concat(..., %unq, ...) -> tensor<?xTy1>
+// clang-format on
+struct CheckUnsqueezeOp : public OpRewritePattern<mlir::ONNXUnsqueezeOp> {
```

Comment:
Is it better to check if the operands have the same type as the result of concat? If not, then you add a cast to the operand. This shouldn't be specific to Unsqueeze?

Reply:
Check ConcatOp directly seems better